Python/Product/PythonTools/PythonTools/Intellisense/AnalysisEntryService.cs

```diff
@@ -0,0 +1,198 @@
+        /// example if a buffer is outside of any projects, but hosted in a difference view with a buffer
+        /// that is in a project, then we'll look in the view that has the project.
+        /// 
+        /// For interactive windows we will use the analyzer that's configured for the window.
+        /// </summary>
+        public bool TryGetAnalysisEntry(ITextView textView, ITextBuffer textBuffer, out AnalysisEntry entry) {
+            ProjectAnalyzer analyzer;
+            string filename;
+            if ((textView == null || !TryGetAnalyzer(textView, out analyzer, out filename)) &&
+                !TryGetAnalyzer(textBuffer, out analyzer, out filename)) {
+                entry = null;
+                return false;
+            }
+
+            var vsAnalyzer = analyzer as VsProjectAnalyzer;
+            if (vsAnalyzer == null) {
+                entry = null;
+                return false;
+            }
+
+            entry = vsAnalyzer.GetAnalysisEntryFromPath(filename);
+            return entry != null;
+        }
+
+        /// <summary>
+        /// Gets the analysis entry for the given buffer.
+        /// 
+        /// This will only succeed if the buffer is a file on disk. It is not able to support things like
+        /// difference views because we don't know what view this buffer is hosted in. This method should
+        /// only be used when we don't know the current view for the buffer. Otherwise, use
+        /// <see cref="TryGetAnalysisEntry(ITextView, ITextBuffer, out AnalysisEntry)"/>
+        /// </summary>
+        public bool TryGetAnalysisEntry(ITextBuffer textBuffer, out AnalysisEntry entry) {
+            return TryGetAnalysisEntry(null, textBuffer, out entry);
+        }
+
+        /// <summary>
+        /// Gets the internal analyzer object from either a text view or buffer. Both arguments
+        /// are optional, and <c>null</c> may be passed.
+        /// </summary>
+        public VsProjectAnalyzer GetVsAnalyzer(ITextView view, ITextBuffer buffer) {
+            VsProjectAnalyzer vsAnalyzer;
+            ProjectAnalyzer analyzer;
+            string filename;
+            if (buffer != null && TryGetAnalyzer(buffer, out analyzer, out filename) &&
+                (vsAnalyzer = analyzer as VsProjectAnalyzer) != null) {
+                return vsAnalyzer;
+            }
+            if (view != null && TryGetAnalyzer(view, out analyzer, out filename) &&
+                (vsAnalyzer = analyzer as VsProjectAnalyzer) != null) {
+                return vsAnalyzer;
+            }
+            return null;
+        }
+
+        #region IAnalysisEntryService members
+
+        public ProjectAnalyzer DefaultAnalyzer => _site.GetPythonToolsService().DefaultAnalyzer;
+
+        public Task WaitForAnalyzerAsync(ITextBuffer textBuffer, CancellationToken cancellationToken) {
+            if (textBuffer.Properties.ContainsProperty(typeof(VsProjectAnalyzer))) {
```

Comment:
>textBuffer [](start = 16, length = 10)

Given this is a public API we are exposing, you should throw ArgumentNullException instead of causing a NRE